### ISSUE TITLE
How to determine whether a Linen Module is still initializing?

### ISSUE THREAD

**Issue by marcvanzee** [google/flax#652]
We currently check whether a module is initialized using `self.has_variable`, for instance in our definition of [MultiHeadDotProductAttention](https://github.com/google/flax/blob/master/flax/linen/attention.py#L200):

```
...
# detect if we're initializing by absence of existing cache data.
is_initialized = self.has_variable('cache', 'cached_key')
...
if is_initialized:
  ...
```

I think this looks a bit strange to the average user. In the old API we were using `self.is_initialized`, and I think we should consider bringing this back in some form or another.

Below some remarks made by different users offline:

From @levskaya : "The reason we did it this way is that it's strictly more general if you imagine having separate variable collections that might have a more complicated initialization than just a single-shot "init" function.  That said, everyone seems to hate it, so I'm not sure the generality is worth not having an `is_initializing` module-global state variable..."

From @jheek : "You need to be careful with the definition of initializing. The simplest one is we called Module.init and not Module.apply. But there are other cases like if I call the same Module in init for the second time it's already fully initialized so should self.is_initializing return True or False? Something like this:

```
def __call__(self, x):
  fc = Dense(5)
  x = fc(x) # this will init kernel, bias...
  x = fc(x) # already initialized. 
```

So to summarize: it seems useful to be able to have a boolean in Module specifying whether a module is initialized or not, but there are some tricky cases we should think about (see example above).

**Comment by jheek** [comment]
I'm not sure if cache is the best example to frame this discussion around. Ideally cache doesn't exist up until decoding stage. Right now we use a slightly odd workaround because we cannot init carry variables inside a loop (but I think we can fix that).

**Comment by cgarciae** [comment]
In #2052 I had to use something like:

```python
is_initializing = self.scope.is_collection_empty("batch_stats")
```
the module itself didn't have variables, initialization state depended on its sub-modules. This solution was very tricky (had to look for option in the Scope API), wondering if a method that checks if the module contain any variables at all would be easier to use and applicable in most common cases:

```
def variables_empty(self) -> bool:
  return bool(self.scope._variables)
```
Alternatively `is_collection_empty` could be added exposed as a method in `Module`.

**Comment by cgarciae** [comment]
@marksandler2 I was looking into this, `BatchNorm` defines initialization like this:

```python
initializing = self.is_mutable_collection('params')
```
Which results in batch statistics not being updated because of:
```python
if not initializing:
  ra_mean.value = self.momentum * ra_mean.value + (self.momentum) * mean
  ra_var.value = self.momentum * ra_var.value + (1 - self.momentum) * var
```
@jheek from the outside it's not clear why making `params` mutable results for `BatchNorm` not training properly. If we want to avoid this issue maybe additional context information about whether we are inside `init` or not is necessary.

**Comment by marksandler2** [comment]
Yes, it is obvious why it is not doing things properly. It is just  a very
sharp edge in flax (and not necessarily batchnorm specific). Re batch norm
specifically this could be fixed by checking if batch_stats 'mean' exists
instead. But overall i think it should be fixed at the API level.


Potentially i think the problem is that today `mutable` conflates two
different operations: "init" and "update", it would be nice if those were
separate then we can make variable updateable, but guarantee that we won't
by accident run "init" on it, and vice-versa.

**Comment by cgarciae** [comment]
> batch norm specifically this could be fixed by checking if batch_stats 'mean' exists instead

This was my first though but it has an edge case, which also happens to be an edge case for what I just proposed: if you have some pretrained parameters but want to initialize others currently it can be solved like this e.g:

```python
variables = FrozenDict({
  "params: {"backbone": pretrained["params"]}, 
  "batch_stats": {"backbone": pretrained["batch_stats"]}
})

# use apply to initialize missing
_, init_updates = module.apply(variables, ..., rng={"params": ...}, mutable=True)
variables = variables.copy(init_updates)
```
Meaning you don't want a variable existing imply the Module is initialized (could slightly deteriorate your batch statistics if you pass a fake sample). Also, `init` currently doesn't optionally accept preexisting variables so my proposal wouldn't work.

Possible solution:
* Create some property that lets you distinguish between an `init` vs `apply` call.
* Let `init` optionally accept pre-existing `variables`.

**Comment by marksandler2** [comment]
I think the edge case is resolved pretty nicely if we update based on the
variable existing or not.

If a variable exists, then it should be updated, if the variable doesn't
exist then it should be initialized, but *not* updated.
Of course in practice checking for each variable independently is messy and
error prone, so the shortcut would be to check
a single variable and that could lead to invalid updates.

**Comment by cgarciae** [comment]
To summarize: the entanglement between mutable state of `params` and initialization logic in `BatchNorm` silently brings issues, I have been bitten by this at least 2 times and its very confusing until you read the source code. It would be nice to solve this somehow.

**Comment by marcvanzee** [comment]
Assinging this to @jheek for now since he has been thinking about this, but we will discuss this offline as well and might reassign.

**Comment by lkhphuc** [comment]
> From @jheek : "You need to be careful with the definition of initializing. The simplest one is we called Module.init and not Module.apply. But there are other cases like if I call the same Module in init for the second time it's already fully initialized so should self.is_initializing return True or False?

Naive question, but is it not that a `self.is_initializing == True` means the module is called from inside the `module.init(rng, inputs, ...)` call, and `self.is_initializing == False` means it's called from `module.apply(...)`.

Then I think the behavior of `self.is_initializing` and `self.has_variable(...)` is orthogonal. One checks the `init()` or `apply()` calls, the other check the structure of the (passed in) variable dict. Currently most modules are implicit defining a `is_initializing` based on a module's specific assumption about the difference in variables dict between apply and init.

**Comment by marksandler2** [comment]
I don't believe is_initializing is a thing in `flax.linen`

**Comment by levskaya** [comment]
#2234 adds an explicit check method `self.is_initializing` for the use of `nn.init(...)(...)` or `module.init(...)`.

**Comment by levskaya** [comment]
closing this issue for now.
